src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -5166,13 +5166,131 @@ int RGWRados::check_bucket_empty(const DoutPrefixProvider *dpp, RGWBucketInfo& b
+        r = ret;
+        break;
+      }
+    }
+
+  } while (r == -ECANCELED && ++retries < max_retries);
+
+  if (r == 0) {
+    for (int i = 0; i < shards_num; ++i) {
+      ldpp_dout(dpp, 10) << "adding to data_log shard_id: " << i << " of gen:" << index_log.gen << dendl;
+      int ret = svc.datalog_rados->add_entry(dpp, bucket_info, index_log, i,
+                                                  null_yield);
+      if (ret < 0) {
+        ldpp_dout(dpp, 1) << "WARNING: failed writing data log for bucket="
+        << bucket_info.bucket << ", shard_id=" << i << "of generation="
+        << index_log.gen << dendl;
+        } // datalog error is not fatal
+  }
+  }
+
+  return r;
+}
+
+int get_zone_ids(const DoutPrefixProvider *dpp,
+				        rgw::sal::RadosStore* const driver,
+                const rgw_zone_id source_zone,
+                const rgw_bucket& bucket,
+                std::vector<rgw_zone_id>& zids,
+                optional_yield y)
+{
+  RGWBucketSyncPolicyHandlerRef source_handler;
+  int ret = driver->get_sync_policy_handler(dpp, source_zone, bucket, &source_handler, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, 10) << "failed to get bucket sync policy handler (r=" << ret << ")" << dendl;
+    return ret;
+  }
+
+  auto all_dests = source_handler->get_all_dests();
+  rgw_zone_id last_zid;
+  for (auto& diter : all_dests) {
+    const auto& zid = diter.first;
+    if (zid == last_zid) {
+      continue;
+    }
+    last_zid = zid;
+    zids.push_back(zid);
+  }
+  return 0;
+}
+
+int list_remote_buckets(const DoutPrefixProvider *dpp,
+				               rgw::sal::RadosStore* const driver,
+                       const rgw_zone_id source_zone,
+                       const rgw_bucket& bucket,
+                       optional_yield y)
+{
+
+  std::vector<rgw_zone_id> zids;
+  int ret = get_zone_ids(dpp, driver, source_zone, bucket, zids, y);
+  if (ret < 0) {
+    ldpp_dout(dpp, 10) << "failed to get remote zones (r=" << ret << ")" << dendl;
+    return ret;
+  }
+
+  std::vector<bucket_unordered_list_result> peer_status;
+  peer_status.resize(zids.size());
+
+  RGWCoroutinesManager crs(driver->ctx(), driver->getRados()->get_cr_registry());
+  RGWHTTPManager http(driver->ctx(), crs.get_completion_mgr());
+  ret = http.start();
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "failed in http_manager.start() ret=" << ret << dendl;
+    return ret;
+  }
+
+  ret = crs.run(dpp, new RGWStatRemoteBucketCR(dpp, driver, source_zone, bucket, &http, zids, peer_status));
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "failed to fetch remote bucket stats " << cpp_strerror(ret) << dendl;
+    return ret;
+  }
+
+  for (const auto& list_result: peer_status) {
+    auto entries_iter = list_result.entries.begin();
+    for (; entries_iter != list_result.entries.end(); ++entries_iter) {
+      std::string ns;
+      rgw_obj_key obj;
+
+      if (rgw_obj_key::oid_to_key_in_ns(entries_iter->key.name, &obj, ns)) {
+        return -ENOTEMPTY;
```

Comment:
@cbodley how about just parsing the oid instead? I have added that in commit 3d2b94f8da54e762b2a9d88e8bf268ce462e00ad do you see any problems with that?

Reply:
sorry, i think we're confusing the two different layers of bucket listings. the ones we request directly from rados are unfiltered, so may contain hidden/namespaced entries (like the parts of incomplete multipart uploads that haven't turned into real objects yet). when we request listings from the S3 API, those hidden entries do get filtered out of the response

since we're fetching the S3 listings with RGWStatRemoteBucketCR, we shouldn't need to do any extra filtering here. we should just be able to check if `!list_result.entries.empty()` for the ENOTEMPTY case

Reply:
@cbodley fixed. thanks! code is complete now, hopefully. will run it through teuthology and report soon.